go-controller/pkg/util/gateway-init.go

```diff
@@ -39,68 +39,80 @@ func getLocalSystemID() (string, error) {
-			"stdout: %q, stderr: %q, error: %v", stdout, stderr, err)
+func ensureGatewayPortAddress(portName string) (string, *net.IPNet, error) {
+	routerMac, routerIP, _ := GetPortAddresses(portName, false)
+	if routerMac == "" || routerIP == "" {
+		// Create the gateway switch port in 'join' if it doesn't exist yet
+		stdout, stderr, err := RunOVNNbctl("--wait=sb",
+			"--may-exist", "lsp-add", "join", portName,
+			"--", "--if-exists", "clear", "logical_switch_port", portName, "dynamic_addresses",
+			"--", "lsp-set-addresses", portName, "dynamic")
+		if err != nil {
+			return "", nil, fmt.Errorf("failed to add logical switch "+
+				"port %s, stdout: %q, stderr: %q, error: %v",
+				portName, stdout, stderr, err)
+		}
+		// Wait for its addresses
+		routerMac, routerIP, err = GetPortAddresses(portName, false)
+		if err != nil {
+			return "", nil, fmt.Errorf("Error while waiting for addresses "+
+				"for gateway switch port %q: %v", portName, err)
+		}
 	}
-}
 
-func generateGatewayIP() (string, error) {
-	// All the routers connected to "join" switch are in 100.64.1.0/24
-	// network and they have their external_ids:connect_to_join set.
-	stdout, stderr, err := RunOVNNbctl("--data=bare", "--no-heading",
-		"--columns=network", "find", "logical_router_port",
-		"external_ids:connect_to_join=yes")
+	// Grab the 'join' switch prefix to add to our gateway router's IP
+	joinSubnet, stderr, err := RunOVNNbctl("--if-exists", "get",
+		"logical_switch", "join", "other-config:subnet")
 	if err != nil {
-		logrus.Errorf("Failed to get logical router ports which connect to "+
-			"\"join\" switch, stdout: %q, stderr: %q, error: %v",
-			stdout, stderr, err)
-		return "", err
+		return "", nil, fmt.Errorf("Failed to get 'join' switch subnet: "+
+			"%s, stderr: %q, %v", joinSubnet, stderr, err)
 	}
-	// Convert \r\n to \n to support Windows line endings
-	stdout = strings.Replace(strings.TrimSpace(stdout), "\r\n", "\n", -1)
-	ips := strings.Split(stdout, "\n")
-
-	ipStart, ipStartNet, _ := net.ParseCIDR("100.64.1.0/24")
-	ipMax, _, _ := net.ParseCIDR("100.64.1.255/24")
-	n, _ := ipStartNet.Mask.Size()
-	for !ipStart.Equal(ipMax) {
-		ipStart = NextIP(ipStart)
-		used := 0
-		for _, v := range ips {
-			ipCompare, _, _ := net.ParseCIDR(v)
-			if ipStart.String() == ipCompare.String() {
-				used = 1
-				break
-			}
-		}
-		if used == 1 {
-			continue
-		} else {
-			break
-		}
+	if joinSubnet == "" {
+		return "", nil, fmt.Errorf("Empty 'join' switch subnet")
 	}
-	ipMask := fmt.Sprintf("%s/%d", ipStart.String(), n)
-	return ipMask, nil
+	joinSubnetSplit := strings.Split(strings.Trim(joinSubnet, `"`), "/")
+	if len(joinSubnetSplit) != 2 {
+		return "", nil, fmt.Errorf("Invalid 'join' switch subnet %q", joinSubnet)
+	}
+	stringCIDR := routerIP + "/" + joinSubnetSplit[1]
+	ip, cidr, err := net.ParseCIDR(stringCIDR)
+	if err != nil {
+		return "", nil, fmt.Errorf("Invalid router CIDR %q: %v", stringCIDR, err)
+	}
+	cidr.IP = ip
+
```

Comment:
Should we stash the router ip address for the join switch subnet in the join switch itself as an external_id? It is not going to change for the duration of the cluster, s we could benefit from not computing it everytime a K8s node is added? Thanks.

Reply:
@girishmg it's actually only using the prefix length of the 'join' switch (eg /24) and adding that to the gateway logical switch port's dynamic address so we get a full CIDR address. I updated the code to stash the prefix length as other-config on the 'join' switch.